freeipa/src/main/java/com/sequenceiq/freeipa/service/image/ImageService.java

```diff
@@ -53,46 +65,111 @@ public Image create(Stack stack, ImageSettingsRequest imageRequest) {
-        if (image.isEmpty()) {
-            imageCatalogProvider.evictImageCatalogCache(catalogUrl);
-            images = imageCatalogProvider.getImageCatalog(catalogUrl).getImages().getFreeipaImages();
-            image = findImage(imageId, imageOs, images, region, platform);
-            if (image.isEmpty()) {
-                throw new ImageNotFoundException(String.format("Could not find any image with id: '%s' in region '%s' with OS '%s'.", imageId, region, imageOs));
-            }
+        ImageCatalog cachedImageCatalog = imageCatalogProvider.getImageCatalog(catalogUrl);
+        return findImageForAppVersion(region, platform, imageId, imageOs, cachedImageCatalog)
+                .or(() -> {
+                    LOGGER.debug("Image not found with the parameters: imageId: {}, imageOs: {}, region: {}, platform: {}", imageId, imageOs, region, platform);
+                    LOGGER.debug("Evicting image catalog cache to retry.");
+                    imageCatalogProvider.evictImageCatalogCache(catalogUrl);
+                    ImageCatalog renewedImageCatalog = imageCatalogProvider.getImageCatalog(catalogUrl);
+                    return findImageForAppVersion(region, platform, imageId, imageOs, renewedImageCatalog);
+                })
+                .orElseThrow(() -> {
+                    String message = String.format("Could not find any image with id: '%s' in region '%s' with OS '%s'.", imageId, region, imageOs);
+                    return new ImageNotFoundException(message);
+                });
+    }
+
+    private Optional<Image> findImageForAppVersion(String region, String platform, String imageId, String imageOs, ImageCatalog catalog) {
+        List<FreeIpaVersions> versions = filterFreeIpaVersionsByAppVersion(catalog.getVersions().getFreeIpaVersions());
+        List<Image> compatibleImages = findImage(imageId, imageOs, catalog.getImages().getFreeipaImages(), region, platform);
+        LOGGER.debug("[{}] compatible images found, by the following parameters: imageId: {}, imageOs: {}, region: {}, platform: {}",
+                compatibleImages.size(), imageId, imageOs, region, platform);
+
+        LOGGER.debug("Attempt to find a default image to use.");
+        return filterImages(versions, compatibleImages, FreeIpaVersions::getDefaults).or(
+                () -> {
+                    LOGGER.debug("Default image not found. Attempt to find an image, compatible with the application version.");
+                    return filterImages(versions, compatibleImages, FreeIpaVersions::getImageIds);
+                }
+        ).or(
+                () -> {
+                    LOGGER.debug("Not found any image compatible with the application version. Falling back to the most recent image.");
+                    return compatibleImages.stream().max(Comparator.comparing(Image::getDate));
+                }
```

Comment:
these should be moved to a well named method so I don't have to check the log lines to see whats happening :)